src/main/java/io/appium/java_client/remote/AppiumCommandExecutor.java

```diff
@@ -60,45 +65,97 @@ public class AppiumCommandExecutor extends HttpCommandExecutor {
 
     private final Optional<DriverService> serviceOptional;
 
-    private AppiumCommandExecutor(Map<String, CommandInfo> additionalCommands, DriverService service,
-                                  URL addressOfRemoteServer,
-                                  HttpClient.Factory httpClientFactory,
-                                  ClientConfig clientConfig) {
+    private final HttpClient.Factory httpClientFactory;
+
+    private final AppiumClientConfig appiumClientConfig;
+
+
+    /**
+     * Create an AppiumCommandExecutor instance.
+     *
+     * @param additionalCommands is the map of Appium commands
+     * @param service take a look at {@link DriverService}
+     * @param addressOfRemoteServer is the address of remotely/locally started Appium server
+     * @param httpClientFactory take a look at {@link HttpClient.Factory}
+     * @param appiumClientConfig take a look at {@link AppiumClientConfig}
+     */
+    public AppiumCommandExecutor(
+            @Nonnull Map<String, CommandInfo> additionalCommands,
+            @Nullable DriverService service,
+            @Nullable URL addressOfRemoteServer,
+            @Nullable HttpClient.Factory httpClientFactory,
+            @Nullable AppiumClientConfig appiumClientConfig) {
+
         super(additionalCommands,
-                ofNullable(clientConfig).orElse(
-                        ClientConfig.defaultConfig()
-                                .baseUrl(Require.nonNull("Server URL", ofNullable(service)
+                ofNullable(appiumClientConfig)
+                        .orElse(AppiumClientConfig.defaultConfig())
```

Comment:
Please use orElseGet if the argument is not a constant.
By using orElse an unnecessary instance is created even though it might not be used.